### ISSUE TITLE
Foldable and Traversable instances

### ISSUE THREAD

**Issue by ekmett** [ixmatus/prizm#1]
I had a use for this package today, but had to use standalone deriving to produce `Foldable` and `Traversable` instances for the CIELAB and RGB types in particular.

Since you've already gone and put in `Applicative` and `Functor` you may want to add those instances.

Also, there is a legal `Monad` for them. e.g.

```haskell
CIELAB x y z >>= f = CIELAB x' y' z' where
  CIELAB x' _ _ = f x
  CIELAB _ y' _ = f y
  CIELAB _ _ z' = f z
```

And `MonadFix`, `MonadZip` instances are also law abiding if you want to go overboard. :)

Similar instances apply to the other colorspace types.

**Comment by ixmatus** [comment]
I just saw this, as I've been updating and modernizing the library. Thanks for the suggestions!

I had functor and applicative instances that necessitated the RGB and CIE* types be of kind `* -> *` which seemed inconvenient to me. I wanted to make the types of kind `*` in order to simplify their use so I had moved to newtype wrapped versions but that seemed even more awkward. I ended up finding the `mono-traversable` library but curious if you have other or better suggestions for what I want to do here.

**Comment by ixmatus** [comment]
That probably was a bit vague around the newtype part. I left the newtype approach and created types of kind `*` with mono-traversable instances but since you're a user of this, is that more inconvenient then what I had before given that you could write traversable and foldable instances?

**Comment by ixmatus** [comment]
Looking at this a bit more, I can at least write the mono-foldable and mono-traversable instances too :)

**Comment by ixmatus** [comment]
~~And now that you have me thinking about it, a Monoid instance would be really nice too.~~ Hmm, that might be difficult because what is a "mempty" color? Hmmm...

**Comment by ixmatus** [comment]
~~Or, better yet, I can provide a top-level `Color a` type that is polymorphic while keeping the individual color types monomorphic. Oooh, this is giving me a lot of ideas; I'll stop talking out-loud in the issue ticket here, but thanks :)~~

**Comment by ixmatus** [comment]
And derp for me, that wouldn't work either because the type variable _must_ be polymorphic for `Functor`.

**Comment by ixmatus** [comment]
Hmm, i could make a color form a monoid if I add an opacity value to a color, though.

**Comment by ekmett** [comment]
There is merit to both the * -> * approach you've taken now and to the monomorphic versions that tie you to some specific data type. Had you gone with the latter, I probably wouldn't have been able to use your code in my application, but I can understand the appeal of a low cost representation. That said, if you're going to say with * -> *, it might be worth it to go all the way to get all the power of the instances for your users.

**Comment by ekmett** [comment]
As for the monoid, the problem is the selection of monoid is messy. I'm partial to the premultiplied alpha monoid, myself. =)

**Comment by ixmatus** [comment]
I think I was going in the right direction before with the newtypes except I had the types flipped around, which resulted in a lot of awkwardness; I think I have something now that enables `Functor`, `Applicative`, etc. instances. e.g: `newtype ColorCoord a = ColorCoord (a,a,a)` and then `newtype RGB = RGB (ColorCoord Word8)`...

Coincidental you just commented :) Thanks for responding and the feedback, I think I'm going to return to types with a kind signature of `* -> *`. And yes, I think I'm going to add a monoidal instance for an alpha channel RGB but it would only work on RGB not the other color spaces.

**Comment by ixmatus** [comment]
I've addressed this with #6. Thanks for the feedback @ekmett!
